### DISCUSSION TITLE
how to generate migrations

### DISCUSSION THREAD

**Discussion by BrianHicks** [automerge/automerge#871] [Q&A]
[The docs](https://automerge.org/docs/cookbook/modeling-data/#versioning) say that you should generate and hardcode an initial migration, and then to "hard-code migrations that upgrade from one schema version to the next, using the same technique (either hard-coding the change as a byte array, or making a change on the fly with hard-coded actorId and timestamp)."

Problem is, I can't figure out the right way to do either of those options! As far as I can see, `Automerge.change` does not allow you to set the actor ID in the change options (or anyway it's not in the TS types.) I've tried this to migrate from `SchemaV1` to `SchemaV2`:

```typescript
type PingV1 = {
  time: Date
  tag: Array<string>
}

type SettingsV1 = {
  lambda: number
}

type SchemaV1 = {
  pings: PingV1[]
  version: 1
}

type SchemaV2 = {
  pings: PingV1[]
  settings: SettingsV1
  version: 2
}

// use a consistent actor ID to migrate
const migrator = "00"

const initialDocument = Automerge.save(
  Automerge.change(Automerge.init<SchemaV1>({ actor: migrator }), (doc) => {
    doc.pings = []
    doc.version = 1
  }),
)

const migrateV1toV2 = Automerge.save(
  Automerge.change<SchemaV2>(
    Automerge.load(initialDocument, { actor: migrator }),
    {
      message: "migrate v1 to v2",
      time: new Date(2024, 2 /* March */, 13, 0, 0, 0, 0).getTime(),
    },
    (doc) => {
      doc.settings = { lambda: 1 / 45 }
      doc.version = 2
    },
  ),
)

```

However, when I try to migrate the doc:

```typescript
let doc = Automerge.load(initialDocument)
Automerge.applyChanges(doc, [migrateV1toV2])
```

I get an error that looks like this:

```
Uncaught (in promise) RangeError: error loading change 0: wrong chunk type
    at __wbg_new_9fa4bec794556db1 (automerge_wasm_bg.js:1870:17)
    at js_sys::RangeError::new::h173ff388a8be4338 (automerge_wasm_bg-UGHCZMDS.wasm:0x14fd70)
    at automerge_wasm::error::<impl core::convert::From<automerge_wasm::error::ApplyChangesError> for wasm_bindgen::JsValue>::from::h6e6432f0e2a5d8a1 (automerge_wasm_bg-UGHCZMDS.wasm:0x12bd6f)
    at automerge_applyChanges (automerge_wasm_bg-UGHCZMDS.wasm:0x140b92)
    at _Automerge.applyChanges (automerge_wasm_bg.js:1110:18)
    at applyChanges (stable.js:686:18)
    at Module.start3 (repo.ts:70:24)
    at async main (index.ts:12:16)
```

I get a similar error when trying to do this with automerge-repo. Truth be told, I'm a little confused about how this should actually work—is calling `Automerge.applyChanges` on the result of `await handle.doc()` the right thing to do?

```typescript
export async function start(): Promise<DocHandle<SchemaV2>> {
  const repo = new Repo({
    network: [new BroadcastChannelNetworkAdapter()],
    storage: new IndexedDBStorageAdapter(),
  })

  let rootDocId = localStorage.rootDocId
  let handle: DocHandle<SchemaV1>

  if (!rootDocId) {
    handle = repo.import<SchemaV1>(initialDocument)
    localStorage.rootDocId = rootDocId = handle.documentId
  } else {
    handle = repo.find(rootDocId)
  }

  let doc = await handle.doc()
  if (doc.version === 1) {
    ;[doc] = Automerge.applyChanges(doc, [migrateV1toV2])
  }

  return handle
}
```

Is it obvious to anyone what I'm doing wrong here? I don't feel confident that I'm using the APIs correctly anyway—for example, I don't see a way to use a consistent actor ID without loading the document fresh. Does that actually work in this case? Or is there something else going on?

**Edit:** I also ought to mention that this happens on both the stable and next versions of Automerge!

**Comment by alexjg** [comment]
I think the problem here is that `Automerge.applyChanges` expects to be passed the output of `Automerge.saveIncremental` rather than the output of `save` (this is a confusing API which we will revamp in the next major). I think what you want to do is `Automerge.loadIncremental(doc, migrateV1toV2)`

**Reply by BrianHicks** [reply]
Ah, fantastic! That worked!

However, I still don't understand how to do this with `automerge-repo`. If I have this code:

```typescript
export async function start(): Promise<DocHandle<SchemaV2>> {
  const repo = new Repo({
    network: [new BroadcastChannelNetworkAdapter()],
    storage: new IndexedDBStorageAdapter(),
  })

  let rootDocId = localStorage.rootDocId
  let handle: DocHandle<SchemaV1>

  if (!rootDocId) {
    handle = repo.import<SchemaV1>(initialDocument)
    localStorage.rootDocId = rootDocId = handle.documentId
  } else {
    handle = repo.find(rootDocId)
  }

  let doc = await handle.doc()
  if (doc.version === 1) {
    doc = Automerge.loadIncremental(doc, migrateV1toV2)
  }

  return handle
}
```

It doesn't look like the change to `doc` is reflected in the document in `handle`. Do I need to do something like re-importing the document after migrations?

**Reply by BrianHicks** [reply]
I dove into the code and found `DocHandle.update`, which seems to work OK. Here's how my `start` function looks now:

```typescript
export async function start(): Promise<DocHandle<SchemaV2>> {
  const repo = new Repo({
    network: [new BroadcastChannelNetworkAdapter()],
    storage: new IndexedDBStorageAdapter(),
  })

  let rootDocId = localStorage.rootDocId
  let handle: DocHandle<{ version: number }>

  if (!rootDocId) {
    handle = repo.import(initialDocument)
    localStorage.rootDocId = rootDocId = handle.documentId
    console.log("created new doc")
  } else {
    handle = repo.find(rootDocId)
    console.log("found existing doc")
  }

  await handle.whenReady()

  handle.update((doc: { version: number }) => {
    if (doc.version === 1) {
      console.log("migrating v1 -> v2")
      doc = Automerge.loadIncremental(doc, migrateV1toV2)
    }

    return Automerge.clone(doc)
  })

  console.log(await handle.doc())

  return handle
}
```

Is this the right pattern? I struggled enough with this to think it might be a useful contribution to the docs if it looks like the intended pattern. The doc comment about `update` receiving changes from the network threw me off for a little but it seems to do what I want?

**Reply by alexjg** [reply]
`DocHandle.update` is I think not intended to be a public API. It looks to me like what we might actually need here is to implement something like `DocHandle.loadIncremental` @acurrieclark @pvh does that sound right to you?

**Reply by BrianHicks** [reply]
also I did some thinking, and how is this pattern supposed to handle stuff in lists? Surely we can’t migrate every key at once, right?
